### PR TITLE
API Changes based on Feedback

### LINKED ISSUE #4: Valid lock names are valid container names
Cam locally validate this in code without prior to making remote call

### LINKED ISSUE #6: Allow full name of Azure Storage account to be provided
cc/ @jjcollinge

### PR DESCRIPTION
**Pull request by lawrencegripper**
Fix #6 Fix #4

### CODE REVIEW

File: locking.go [hunk 1]
```diff
@@ -137,28 +135,50 @@ var (
...
+	accountName, err := extractAccountNameFromURL(storageAccountURLParsed)
+	if err != nil {
+		return nil, err
+	}
+
+	creds := azblob.NewSharedKeyCredential(accountName, storageAccountKey)
 
 	// Create a ContainerURL object to a container
-	u, _ := url.Parse(fmt.Sprintf("https://%s.blob.core.windows.net/%s", accountName, lockName))
+	u, _ := url.Parse(fmt.Sprintf("%s/%s", storageAccountURL, lockName))
```
Comment (jjcollinge) [inline locking.go]: Can we add a check to validate no existing path has been set on the URL .i.e. passing in a valid storage URL with the container already on it like so: https://mystorageaccount.blob.core.windows.net/containername would pass the current validation but result in an ill formatted URL like: https://mystorageaccount.blob.core.windows.net/containername/containername

File: locking.go [hunk 2]
```diff
@@ -137,28 +135,50 @@ var (
...
 	containerURL := azblob.NewContainerURL(*u, azblob.NewPipeline(creds, azblob.PipelineOptions{}))
```
Comment (jjcollinge) [inline locking.go]: Should we add some basic pipeline options i.e. maxRetries

File: locking.go [hunk 3]
```diff
@@ -137,28 +135,50 @@ var (
...
-	_, err := containerURL.Create(ctxParent, nil, azblob.PublicAccessNone)
+	_, err = containerURL.Create(ctxParent, nil, azblob.PublicAccessNone)
 
 	// Create will return a '409' response code if the container already exists
 	// we only error on other conditions as it's expected that a lock of this
 	// name may already exist
-	if err != nil && err.(azblob.ResponseError) != nil && err.(azblob.ResponseError).Response().StatusCode != 409 {
+	_, isReponseError := err.(azblob.ResponseError)
+	if err != nil && isReponseError && err.(azblob.ResponseError).Response().StatusCode != 409 {
```
Comment (jjcollinge) [inline locking.go]: Why do the type assertion twice and not just use the result of the line above here?

File: locking.go [hunk 4]
```diff
@@ -185,7 +205,7 @@ func NewLockInstance(ctxParent context.Context, accountName, accountKey, lockNam
 		lockInstance.used = true
```
Comment (jjcollinge) [inline locking.go]: Is this thread safe? Should we use a mutex, channel or atomic uint32?

Comment (lawrencegripper) [reply locking.go]: Yeah this should be done - good point!

Comment (jjcollinge) [reply locking.go]: same for lockAcquired?

File: locking.go (hunk 1 above)
Comment (lawrencegripper) [reply locking.go]: Good call, will add now

File: locking.go (hunk 2 above)
Comment (lawrencegripper) [reply locking.go]: Yes will do

File: locking.go (hunk 3 above)
Comment (lawrencegripper) [reply locking.go]: Merge fail :( Will fix

File: locking.go (hunk 4 above)
Comment (lawrencegripper) [reply locking.go]: Added a mutex

Comment (lawrencegripper) [reply locking.go]: Do you think its right - I now lock on any of the operations so you can't 'Renew' while your already unlocking etc.

Comment (jjcollinge) [reply locking.go]: Looks ok - do you need a test to confirm?

Comment (lawrencegripper) [reply locking.go]: I'll take a crack at a test.... gonna be a fun one to write!

Comment (lawrencegripper) [reply locking.go]: As discussed this test becomes a test of the mutex and is hard to make into something meaningful so will leave out for now.

**Pull request merged**